pkg/client/openstack/kluster/client.go

```diff
@@ -190,94 +202,249 @@ func (c *klusterClient) SetSecurityGroup(sgName, nodeID string) (err error) {
-			tcp = true
-			continue
+	sg := []string{lbGroup.ID}
+	for _, lb := range allLoadbalancers {
+		if strings.HasPrefix(lb.Name, fmt.Sprintf("kube_service_%s", kluster.Spec.Name)) {
+			found := false
+			for _, port := range allPorts {
+				if lb.VipPortID == port.ID {
+					found = true
+					break
+				}
+			}
+			if !found {
+				_, err := ports.Update(c.NetworkClient, lb.VipPortID, ports.UpdateOpts{SecurityGroups: &sg}).Extract()
+				if err != nil {
+					return false, err
+				}
+			}
 		}
+	}
 
-		if icmp && udp && tcp {
-			break
-		}
+	return true, nil
+}
+
+func (c *klusterClient) EnsureKubernikusRulesInSecurityGroup(kluster *v1.Kluster) (created bool, err error) {
+	sgName := kluster.Spec.Openstack.SecurityGroupName
+	page, err := securitygroups.List(c.NetworkClient, securitygroups.ListOpts{Name: sgName}).AllPages()
+	if err != nil {
+		return false, fmt.Errorf("SecurityGroup %v not found: %s", sgName, err)
 	}
 
-	opts := rules.CreateOpts{
-		Direction:      rules.DirIngress,
-		EtherType:      rules.EtherType4,
-		SecGroupID:     groups[0].ID,
-		RemoteIPPrefix: *kluster.Spec.ClusterCIDR,
+	if kluster.ClusterCIDR() == "" {
+		return false, errors.New("Cluster CIDR for kluster not set")
 	}
 
-	if !udp {
-		opts.Protocol = rules.ProtocolUDP
-		_, err := rules.Create(c.NetworkClient, opts).Extract()
-		if err != nil {
-			return false, err
-		}
+	groups, err := securitygroups.ExtractGroups(page)
+	if err != nil {
+		return false, err
 	}
 
-	if !tcp {
-		opts.Protocol = rules.ProtocolTCP
-		_, err := rules.Create(c.NetworkClient, opts).Extract()
-		if err != nil {
-			return false, err
-		}
+	if len(groups) != 1 {
+		return false, fmt.Errorf("More than one SecurityGroup with name %v found", sgName)
 	}
 
-	if !icmp {
-		opts.Protocol = rules.ProtocolICMP
+	apiIP, err := ipForUrl(kluster.Status.Apiserver)
+	if err != nil {
+		return false, fmt.Errorf("Failed to lookup apiserver ip: %w", err)
+	}
+
+	lbGroup, err := getLoadbalancerSecurityGroup(kluster, c.NetworkClient)
+	if err != nil {
+		// do nothing
+	}
+
+	wantedRules := []rules.SecGroupRule{
+		{
+			Direction:     string(rules.DirIngress),
+			EtherType:     string(rules.EtherType4),
+			RemoteGroupID: groups[0].ID,
+			Description:   fmt.Sprintf(`Kubernikus: accept traffic within security group of cluster "%s"`, kluster.Spec.Name),
+		},
+		{
+			Direction:    string(rules.DirEgress),
+			EtherType:    string(rules.EtherType4),
+			Protocol:     string(rules.ProtocolUDP),
+			PortRangeMin: 123,
+			PortRangeMax: 123,
+			Description:  "Kubernikus: allow ntp client traffic",
+		},
+		{
+			Direction:      string(rules.DirEgress),
+			EtherType:      string(rules.EtherType4),
+			Protocol:       string(rules.ProtocolTCP),
+			PortRangeMin:   443,
+			PortRangeMax:   443,
+			RemoteIPPrefix: apiIP.String(),
+			Description:    fmt.Sprintf(`Kubernikus: allow access to apiserver of cluster "%s"`, kluster.Spec.Name),
+		},
+	}
+
+	if lbGroup != nil && lbGroup.ID != "" {
```

Comment:
This can be nil, eg. when there is not enough quota in the project for creating an additional security group. This ensures the other rules even if that's the case.

Reply:
I see, good point but. it looks to me like at the moment at least if creating the security group fails there is an early return and no other security groups are enforced anyways.

Reply:
Where is it returning? I think I'm ignoring the error while getting the security group.

Reply:
Here: https://github.com/sapcc/kubernikus/pull/601/files#diff-b5e03987e634a7e5e1033109b07cba3c54a48db540f4694445c66794899fdabfR255

Reply:
Isn't this only exiting `EnsureLoadbalancerSecurityGroup()`? To me it looks like `EnsureKubernikusRulesInSecurityGroup()` runs anyway.

Reply:
ah yes, sorry I misread the code. Nvm